llvm/lib/SYCLLowerIR/SYCLOptimizeBarriers.cpp

```diff
@@ -0,0 +1,704 @@
+      // TODO: actually allow semantics missmatch for barriers removal for
+      // several cases.
+      if (Last.Semantic != Cur.Semantic)
+        break;
+
+      auto CmpExec = compareScopesWithWeights(Last.ExecScope, Cur.ExecScope);
+      auto CmpMem = compareScopesWithWeights(Last.MemScope, Cur.MemScope);
+      RegionMemScope FenceLast = getBarrierFencedScope(Last.MemScope);
+      RegionMemScope FenceCur = getBarrierFencedScope(Cur.MemScope);
+
+      if (CmpExec == CompareRes::UNKNOWN || CmpMem == CompareRes::UNKNOWN ||
+          FenceLast == RegionMemScope::Unknown ||
+          FenceCur == RegionMemScope::Unknown)
+        break;
+
+      // If identical then drop Cur.
+      if (CmpExec == CompareRes::EQUAL && CmpMem == CompareRes::EQUAL) {
+        if (noFencedMemAccessesBetween(Last.CI, Cur.CI, FenceLast, BBMemInfo)) {
+          Changed |= eraseBarrierWithITT(Cur);
+        }
+        break;
+      }
+      // If Last wider then drop Cur.
+      if ((CmpExec == CompareRes::BIGGER || CmpMem == CompareRes::BIGGER) &&
+          noFencedMemAccessesBetween(Last.CI, Cur.CI, FenceCur, BBMemInfo)) {
+        Changed |= eraseBarrierWithITT(Cur);
+        break;
+      }
+      // If Cur wider then drop Last and retry.
+      if ((CmpExec == CompareRes::SMALLER || CmpMem == CompareRes::SMALLER) &&
+          noFencedMemAccessesBetween(Last.CI, Cur.CI, FenceLast, BBMemInfo)) {
+        Changed |= eraseBarrierWithITT(Last);
+        Survivors.pop_back();
+        continue;
+      }
+      // No elimination possible.
+      break;
+    }
+    if (Cur.CI) // still alive?
+      Survivors.push_back(Cur);
+  }
+
+  // If we removed any, replace Barriers with the survivors
+  if (Survivors.size() != Barriers.size()) {
+    Barriers.clear();
+    Barriers.append(Survivors.begin(), Survivors.end());
+    Changed = true;
+  }
+  return Changed;
+}
+
+// Remove barriers that are redundant in the CFG based on dominance relations.
+static bool eliminateDominatedBarriers(SmallVectorImpl<BarrierDesc *> &Barriers,
+                                       DominatorTree &DT,
+                                       PostDominatorTree &PDT,
+                                       BBMemInfoMap &BBMemInfo) {
+  bool Changed = false;
+  for (auto *B1 : Barriers) {
+    if (!B1->CI)
+      continue;
+    for (auto *B2 : Barriers) {
+      // Check if the barrier was already removed.
+      if (B1 == B2 || !B2->CI)
+        continue;
+
+      // Skip barriers with missmatching Semantic, Scopes or Unknown Scope.
+      if (B1->Semantic != B2->Semantic)
+        continue;
+      if (B1->ExecScope != B2->ExecScope || B1->MemScope != B2->MemScope)
+        continue;
+      if (B1->ExecScope == Scope::Unknown || B1->MemScope == Scope::Unknown)
+        continue;
+
+      RegionMemScope Fence = getBarrierFencedScope(B1->MemScope);
+      if (Fence == RegionMemScope::Unknown)
+        continue;
+
+      if (DT.dominates(B1->CI, B2->CI)) {
```

Comment:
is there repeated calculation for the case
* B1 = A0, B2 = A1, A0 dominates A1,                           noFencedAccessesCFG returns false
* B1 = A1, B2 = A0, A1 post-dominates A0,                  noFencedAccessesCFG is called again on the same instructions

Reply:
Should be partially resolved.